src/Microsoft.AspNetCore.HttpsPolicy/HttpsRedirectionMiddleware.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _config = config ?? throw new ArgumentException(nameof(config));
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            _serverAddressesFeature = serverAddressesFeature ?? throw new ArgumentNullException(nameof(serverAddressesFeature));
+
+            var httpsRedirectionOptions = options.Value;
+            _httpsPort = httpsRedirectionOptions.HttpsPort;
+            _statusCode = httpsRedirectionOptions.RedirectStatusCode;
+        }
+
+        /// <summary>
+        /// Invokes the HttpsRedirectionMiddleware
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            if (context.Request.IsHttps)
+            {
+                return _next(context);
+            }
+
+            if (!_evaluatedServerAddressesFeature && !_httpsPort.HasValue)
+            {
+                CheckForHttpsPorts();
+            }
+            _evaluatedServerAddressesFeature = true;
+
+            var host = context.Request.Host;
+            if (_httpsPort.HasValue)
+            {
+                host = new HostString(host.Host, _httpsPort.Value);
+            }
+            else
+            {
+                host = new HostString(host.Host);
+            }
+
+            var request = context.Request;
+            var redirectUrl = UriHelper.BuildAbsolute(
+                "https", 
+                host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+
+            context.Response.StatusCode = _statusCode;
+            context.Response.Headers[HeaderNames.Location] = redirectUrl;
+
+            return Task.CompletedTask;
+        }
+
+        private void CheckForHttpsPorts()
+        {
+            // The IServerAddressesFeature will not be ready until the middleware is Invoked,
+            // Order for finding the HTTPS port:
+            // 1. Set in the HttpsRedirectionOptions
+            // 2. HTTPS_PORT environment variable
+            // 3. IServerAddressesFeature (checked in HttpsRedirectionMiddleware.Invoke
+            // 4. 443 (or not set)
+
+            if (_httpsPort != null)
+            {
+                return;
+            }
+
+            var configHttpsPort = _config["HTTPS_PORT"];
+            if (!string.IsNullOrEmpty(configHttpsPort)
+                                && int.TryParse(configHttpsPort, out var intHttpsPort))
+            {
+                _httpsPort = intHttpsPort;
+                return;
+            }
+
+            int? httpsPort = null;
+            foreach (var address in _serverAddressesFeature.Addresses)
+            {
+                var serverAddress = ServerAddress.FromUrl(address);
```

Comment:
This needs to come from a sources package. We don't want this middleware to depend on Kestrel. /cc: @Tratcher